<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Align Report &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Relate Report" href="relate.html" />
    <link rel="prev" title="Report Formats" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../manuals/index.html">Manuals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">File Formats</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Report Formats</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Align Report</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#align-report-fields">Align Report: Fields</a></li>
<li class="toctree-l4"><a class="reference internal" href="#align-report-example">Align Report: Example</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate Report</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../data/index.html">Data Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="../meta/index.html">Metadata Formats</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">File Formats</a></li>
          <li class="breadcrumb-item"><a href="index.html">Report Formats</a></li>
      <li class="breadcrumb-item active">Align Report</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/formats/report/align.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="align-report">
<h1>Align Report<a class="headerlink" href="#align-report" title="Permalink to this headline"></a></h1>
<section id="align-report-fields">
<h2>Align Report: Fields<a class="headerlink" href="#align-report-fields" title="Permalink to this headline"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 83%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Data Type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Name of Sample</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>Use demultiplexed mode</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Use paired-end mode</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Phred score encoding in FASTQ and SAM/BAM/CRAM files</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>Run FastQC on the initial and trimmed FASTQ files</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Trim reads with Cutadapt before alignment</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Phred score for read 1 quality trimming with Cutadapt</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Phred score for read 2 quality trimming with Cutadapt</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>5’ adapter for read 1 adapter trimming with Cutadapt</p></td>
<td><p>list[str]</p></td>
</tr>
<tr class="row-odd"><td><p>3’ adapter for read 1 adapter trimming with Cutadapt</p></td>
<td><p>list[str]</p></td>
</tr>
<tr class="row-even"><td><p>5’ adapter for read 2 adapter trimming with Cutadapt</p></td>
<td><p>list[str]</p></td>
</tr>
<tr class="row-odd"><td><p>3’ adapter for read 2 adapter trimming with Cutadapt</p></td>
<td><p>list[str]</p></td>
</tr>
<tr class="row-even"><td><p>Minimum overlap of read and adapter during trimming with Cutadapt</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Error tolerance for adapters during trimming with Cutadapt</p></td>
<td><p>float</p></td>
</tr>
<tr class="row-even"><td><p>Allow indels in adapters during trimming with Cutadapt</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Trim high-quality Gs from 3’ end during trimming with Cutadapt</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Discard reads in which an adapter was found by Cutadapt</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Discard reads in which no adapter was found by Cutadapt</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Minimum length of a read to keep it after trimming with Cutadapt</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Run Bowtie2 in local mode</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Output discordant alignments from Bowtie2</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Attempt to align individual mates of unaligned pairs with Bowtie2</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Treat dovetailed mate pairs as concordant with Bowtie2</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Treat nested mate pairs as concordant with Bowtie2</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-even"><td><p>Minimum score for a valid alignment with Bowtie2</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>Minimum fragment length for valid paired-end alignments with Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>Maximum fragment length for valid paired-end alignments with Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Minimum distance of an indel from the end of a read with Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>Seed length for Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Seed interval for Bowtie2</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>Maximum number of consecutive failed seed extensions with Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Maximum number of re-seeding attempts with Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>Width of padding on alignment matrix (to allow indels) with Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Valid orientations of paired-end mates with Bowtie2</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>Output unaligned reads from Bowtie2 to a FASTQ file</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>Minimum mapping quality to keep an aligned read from Bowtie2</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>Number of reads initially</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>Number of reads after trimming</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>Number of reads after alignment</p></td>
<td><p>dict[str, int]</p></td>
</tr>
<tr class="row-odd"><td><p>Number of reads after filtering</p></td>
<td><p>dict[str, int]</p></td>
</tr>
<tr class="row-even"><td><p>Number of reads aligned by reference</p></td>
<td><p>dict[str, int]</p></td>
</tr>
<tr class="row-odd"><td><p>Branches</p></td>
<td><p>list[str]</p></td>
</tr>
<tr class="row-even"><td><p>Time Began</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>Time Ended</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>Time Taken (minutes)</p></td>
<td><p>float</p></td>
</tr>
<tr class="row-odd"><td><p>Version of SEISMIC-RNA</p></td>
<td><p>str</p></td>
</tr>
</tbody>
</table>
</section>
<section id="align-report-example">
<h2>Align Report: Example<a class="headerlink" href="#align-report-example" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;Name of Sample&quot;</span><span class="p">:</span> <span class="s2">&quot;ldi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Use demultiplexed mode&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Use paired-end mode&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Phred score encoding in FASTQ and SAM/BAM/CRAM files&quot;</span><span class="p">:</span> <span class="mi">33</span><span class="p">,</span>
    <span class="s2">&quot;Run FastQC on the initial and trimmed FASTQ files&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Trim reads with Cutadapt before alignment&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Phred score for read 1 quality trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span>
    <span class="s2">&quot;Phred score for read 2 quality trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span>
    <span class="s2">&quot;5&#39; adapter for read 1 adapter trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;GCTCTTCCGATCT&quot;</span>
    <span class="p">],</span>
    <span class="s2">&quot;3&#39; adapter for read 1 adapter trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;AGATCGGAAGAGC&quot;</span>
    <span class="p">],</span>
    <span class="s2">&quot;5&#39; adapter for read 2 adapter trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;GCTCTTCCGATCT&quot;</span>
    <span class="p">],</span>
    <span class="s2">&quot;3&#39; adapter for read 2 adapter trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;AGATCGGAAGAGC&quot;</span>
    <span class="p">],</span>
    <span class="s2">&quot;Minimum overlap of read and adapter during trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">,</span>
    <span class="s2">&quot;Error tolerance for adapters during trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="s2">&quot;Allow indels in adapters during trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Trim high-quality Gs from 3&#39; end during trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Discard reads in which an adapter was found by Cutadapt&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Discard reads in which no adapter was found by Cutadapt&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Minimum length of a read to keep it after trimming with Cutadapt&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
    <span class="s2">&quot;Run Bowtie2 in local mode&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Output discordant alignments from Bowtie2&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Attempt to align individual mates of unaligned pairs with Bowtie2&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Treat dovetailed mate pairs as concordant with Bowtie2&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
    <span class="s2">&quot;Treat nested mate pairs as concordant with Bowtie2&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Minimum score for a valid alignment with Bowtie2&quot;</span><span class="p">:</span> <span class="s2">&quot;L,1,0.5&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Minimum fragment length for valid paired-end alignments with Bowtie2&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s2">&quot;Maximum fragment length for valid paired-end alignments with Bowtie2&quot;</span><span class="p">:</span> <span class="mi">600</span><span class="p">,</span>
    <span class="s2">&quot;Minimum distance of an indel from the end of a read with Bowtie2&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
    <span class="s2">&quot;Seed length for Bowtie2&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
    <span class="s2">&quot;Seed interval for Bowtie2&quot;</span><span class="p">:</span> <span class="s2">&quot;L,1,0.1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Maximum number of consecutive failed seed extensions with Bowtie2&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
    <span class="s2">&quot;Maximum number of re-seeding attempts with Bowtie2&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
    <span class="s2">&quot;Width of padding on alignment matrix (to allow indels) with Bowtie2&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
    <span class="s2">&quot;Valid orientations of paired-end mates with Bowtie2&quot;</span><span class="p">:</span> <span class="s2">&quot;fr&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Output unaligned reads from Bowtie2 to a FASTQ file&quot;</span><span class="p">:</span> <span class="n">true</span><span class="p">,</span>
    <span class="s2">&quot;Minimum mapping quality to keep an aligned read from Bowtie2&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span>
    <span class="s2">&quot;Number of reads initially&quot;</span><span class="p">:</span> <span class="mi">566520</span><span class="p">,</span>
    <span class="s2">&quot;Number of reads after trimming&quot;</span><span class="p">:</span> <span class="mi">565068</span><span class="p">,</span>
    <span class="s2">&quot;Number of reads after alignment&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;reads, were paired&quot;</span><span class="p">:</span> <span class="mi">565068</span><span class="p">,</span>
        <span class="s2">&quot;reads, were paired, aligned concordantly 0 times&quot;</span><span class="p">:</span> <span class="mi">2613</span><span class="p">,</span>
        <span class="s2">&quot;reads, were paired, aligned concordantly exactly 1 time&quot;</span><span class="p">:</span> <span class="mi">562448</span><span class="p">,</span>
        <span class="s2">&quot;reads, were paired, aligned concordantly &gt;1 times&quot;</span><span class="p">:</span> <span class="mi">7</span>
    <span class="p">},</span>
    <span class="s2">&quot;Number of reads after filtering&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;paired-end, both mates mapped&quot;</span><span class="p">:</span> <span class="mi">562325</span><span class="p">,</span>
        <span class="s2">&quot;paired-end, one mate unmapped&quot;</span><span class="p">:</span> <span class="mi">0</span>
    <span class="p">},</span>
    <span class="s2">&quot;Number of reads aligned by reference&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;sars2_1799&quot;</span><span class="p">:</span> <span class="mi">562325</span>
    <span class="p">},</span>
    <span class="s2">&quot;Branches&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;Time Began&quot;</span><span class="p">:</span> <span class="s2">&quot;2023-11-10 at 20:16:29&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Time Ended&quot;</span><span class="p">:</span> <span class="s2">&quot;2023-11-10 at 20:17:39&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Time Taken (minutes)&quot;</span><span class="p">:</span> <span class="mf">1.16</span><span class="p">,</span>
    <span class="s2">&quot;Version of SEISMIC-RNA&quot;</span><span class="p">:</span> <span class="s2">&quot;0.9.3&quot;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Report Formats" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="relate.html" class="btn btn-neutral float-right" title="Relate Report" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>