<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Cluster: Infer alternative structures by clustering reads’ mutations &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Join: Merge sections (horizontally) from the Mask or Cluster step" href="join.html" />
    <link rel="prev" title="Mask: Define mutations and sections to filter reads and positions" href="mask.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Merge samples (vertically) from the Relate step</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Cluster: Infer alternative structures by clustering reads’ mutations</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#cluster-input-files">Cluster: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cluster-settings">Cluster: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cluster-output-files">Cluster: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cluster-verify-clusters">Cluster: Verify clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cluster-troubleshoot-and-optimize">Cluster: Troubleshoot and optimize</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="join.html">Join: Merge sections (horizontally) from the Mask or Cluster step</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lists.html">List Positions Matching Criteria</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adjclust.html">Add/Delete Orders to/from an Already-Clustered Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Cluster: Infer alternative structures by clustering reads’ mutations</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/cluster.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cluster-infer-alternative-structures-by-clustering-reads-mutations">
<h1>Cluster: Infer alternative structures by clustering reads’ mutations<a class="headerlink" href="#cluster-infer-alternative-structures-by-clustering-reads-mutations" title="Permalink to this headline"></a></h1>
<section id="cluster-input-files">
<h2>Cluster: Input files<a class="headerlink" href="#cluster-input-files" title="Permalink to this headline"></a></h2>
<section id="cluster-input-file-mask-report">
<h3>Cluster input file: Mask report<a class="headerlink" href="#cluster-input-file-mask-report" title="Permalink to this headline"></a></h3>
<p>You can give any number of Mask report files as inputs for the Cluster step.
See <a class="reference internal" href="../inputs.html"><span class="doc">List Input Files</span></a> for ways to list multiple files.</p>
<p>Cluster all masked reads in <code class="docutils literal notranslate"><span class="pre">out</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">cluster</span> <span class="n">out</span>
</pre></div>
</div>
<p>Cluster reads from <code class="docutils literal notranslate"><span class="pre">sample-1</span></code> masked over reference reference <code class="docutils literal notranslate"><span class="pre">ref-1</span></code>,
section <code class="docutils literal notranslate"><span class="pre">abc</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">cluster</span> <span class="n">out</span><span class="o">/</span><span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">mask</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">abc</span>
</pre></div>
</div>
</section>
</section>
<section id="cluster-settings">
<h2>Cluster: Settings<a class="headerlink" href="#cluster-settings" title="Permalink to this headline"></a></h2>
<section id="cluster-setting-maximum-order-number-of-clusters">
<span id="cluster-max"></span><h3>Cluster setting: Maximum order (number of clusters)<a class="headerlink" href="#cluster-setting-maximum-order-number-of-clusters" title="Permalink to this headline"></a></h3>
<p>To infer alternative RNA structures, SEISMIC-RNA uses an optimized version of
our original DREEM algorithm [<a class="reference external" href="https://doi.org/10.1038/s41586-020-2253-5">Tomezsko et al. (2020)</a>], which is a type of
<a class="reference external" href="https://en.wikipedia.org/wiki/Expectation%E2%80%93maximization_algorithm">expectation-maximization</a> (EM).
All EM algorithms need the order of clustering (i.e. number of clusters) to be
prespecified; however, the optimal order is unknown before the algorithm runs,
creating a <a class="reference external" href="https://en.wikipedia.org/wiki/Chicken_or_the_egg">chicken-and-egg problem</a>.</p>
<p>SEISMIC-RNA solves this problem by first running the EM algorithm at order 1,
then order 2, then 3, and so on until one of two limits is reached:</p>
<ul class="simple">
<li><p>The <a class="reference external" href="https://en.wikipedia.org/wiki/Bayesian_information_criterion">Bayesian information criterion</a> (BIC) worsens upon increasing the order.</p></li>
<li><p>The maximum order is reached.
You can set this limit using <code class="docutils literal notranslate"><span class="pre">--max-clusters</span></code> (<code class="docutils literal notranslate"><span class="pre">-k</span></code>).
If you run the entire workflow using <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code> (see <a class="reference internal" href="wf.html"><span class="doc">Workflow: Run all steps</span></a>), then
the maximum order defaults to 0 (which disables clustering).
If you run the Cluster step individually using <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">cluster</span></code>, then the
maxmimum order defaults to 2 (the minimum non-trivial number).</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the BIC score worsens (increases) before reaching the maximum order,
then clustering will stop.
The report (see <a class="reference internal" href="../../formats/report/cluster.html"><span class="doc">Cluster Report</span></a>) records the maximum
order you specified (field “Maximum Number of Clusters”) and the order that
yielded the best BIC (field “Optimal Number of Clusters”), which is always
less than or equal to the maximum order you specified.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you realize after clustering that it would have been better to have run
clustering with a higher/lower maximum order, then you can edit the results
using <code class="docutils literal notranslate"><span class="pre">+addclust</span></code>/<code class="docutils literal notranslate"><span class="pre">+delclust</span></code> (see <a class="reference internal" href="../adjclust.html"><span class="doc">Add/Delete Orders to/from an Already-Clustered Dataset</span></a>).</p>
</div>
</section>
<section id="cluster-setting-expectation-maximization-iterations">
<h3>Cluster setting: Expectation-maximization iterations<a class="headerlink" href="#cluster-setting-expectation-maximization-iterations" title="Permalink to this headline"></a></h3>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Expectation%E2%80%93maximization_algorithm">Expectation-maximization</a> is an iterative algorithm, meaning that it begins by
guessing an initial solution and then calculates progressively better solutions,
halting once successive solutions cease changing, which is called convergence.</p>
<p>You can limit the minimum/maximum number of iterations per number of clusters
using <code class="docutils literal notranslate"><span class="pre">--min-em-iter</span></code> and <code class="docutils literal notranslate"><span class="pre">--max-em-iter</span></code>, respectively.
Generally, as the number of clusters increases, so does the number of iterations
required for convergence.
Thus, to treat different numbers of clusters more fairly, SEISMIC-RNA multiplies
the iteration limits by the number of clusters.
For example, if you use <code class="docutils literal notranslate"><span class="pre">--max-em-iter</span> <span class="pre">300</span></code>, then SEISMIC-RNA will allow up to
600 iterations for 2 clusters, 900 iterations for 3 clusters, and so on.
The exception is for 1 cluster: since all reads go into the same cluster, there
is no need to iterate, so the iteration limit is always the minimum possible, 2.</p>
<p>You can set the threshold for convergence with <code class="docutils literal notranslate"><span class="pre">--em-thresh</span></code> followed by the
minimum difference between log-likelihoods of successive iterations for the
iterations to be considered different.
For example, if you set the threshold to 0.1 with <code class="docutils literal notranslate"><span class="pre">--em-thresh</span> <span class="pre">0.1</span></code>, then if
iterations 38 and 39 had log-likelihoods of -7.28 and -7.17, respectively, then
the algorithm would keep going because their difference in log-likelihood (0.11)
would exceed the threshold; but if iteration 40 had a log-likelihood of -7.08,
then the algorithm would consider itself converged and stop running because the
difference in log-likelihood between iterations 40 and 39 would be 0.09, which
would be below the threshold.</p>
</section>
<section id="cluster-setting-expectation-maximization-runs">
<h3>Cluster setting: Expectation-maximization runs<a class="headerlink" href="#cluster-setting-expectation-maximization-runs" title="Permalink to this headline"></a></h3>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Expectation%E2%80%93maximization_algorithm">Expectation-maximization</a> is guaranteed to return a locally optimal solution,
but there is no guarantee that the solution will be globally optimal.
To improve the odds of finding the global optimum, SEISMIC-RNA runs EM multiple
times (by default, 6 times), each time starting at a different initial guess.
The idea is that if multiple EM runs, initialized randomly, converge on the same
solution, then that solution is probably the global optimum.
You can set the number of independent EM runs using <code class="docutils literal notranslate"><span class="pre">--em-runs</span></code> (<code class="docutils literal notranslate"><span class="pre">-e</span></code>).</p>
</section>
</section>
<section id="cluster-output-files">
<h2>Cluster: Output files<a class="headerlink" href="#cluster-output-files" title="Permalink to this headline"></a></h2>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">OUT/SAMPLE/cluster/REFERENCE/SECTION</span></code>.</p>
<section id="cluster-output-file-batch-of-cluster-memberships">
<h3>Cluster output file: Batch of cluster memberships<a class="headerlink" href="#cluster-output-file-batch-of-cluster-memberships" title="Permalink to this headline"></a></h3>
<p>Each batch of clustered reads contains a <code class="docutils literal notranslate"><span class="pre">ClustBatchIO</span></code> object and is saved to
the file <code class="docutils literal notranslate"><span class="pre">cluster-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <span class="xref std std-doc">../../data/cluster/cluster</span> for details on the data structure.
See <a class="reference internal" href="../../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for more information on brickle files.</p>
</section>
<section id="cluster-output-file-cluster-report">
<h3>Cluster output file: Cluster report<a class="headerlink" href="#cluster-output-file-cluster-report" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA also writes a report file, <code class="docutils literal notranslate"><span class="pre">cluster-report.json</span></code>, that records the
settings you used for running the Cluster step and summarizes the results, such
as the number of clusters, number of iterations, and the BIC scores.
See <a class="reference internal" href="../../formats/report/cluster.html"><span class="doc">Cluster Report</span></a> for more information.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You <strong>must</strong> look at the report file to determine whether your clusters come
from true alternative structures or are just noise and artifacts.
See <a class="reference internal" href="#clust-verify"><span class="std std-ref">Cluster: Verify clusters</span></a> for how to verify that your clusters are real.</p>
</div>
</section>
</section>
<section id="cluster-verify-clusters">
<span id="clust-verify"></span><h2>Cluster: Verify clusters<a class="headerlink" href="#cluster-verify-clusters" title="Permalink to this headline"></a></h2>
<p>You <strong>must</strong> check whether your clusters are real or artifacts.</p>
<p>In your cluster report:</p>
<ul class="simple">
<li><p>The number of clusters that SEISMIC-RNA found is Optimal Number of Clusters.
Several important caveats exist about this number:</p>
<ul>
<li><p>This number can never exceed the Maximum Number of Clusters.
So if you want to know whether an RNA forms <em>N</em> alternative structures, the
results of clustering can provide useful information only if you set the
Maximum Number of Clusters to at least <em>N</em>.</p></li>
<li><p>A “cluster” is as subjective as a “conformational state”: two clusters can
correspond to completely different structures at one extreme and to slightly
different structures at the other.
With more reads comes better ability to distinguish clusters that are more
similar – the same way that, in a study examining differences between two
groups, larger sample sizes would enable finding more subtle differences.
Thus, the number of clusters you find will generally increase with more
reads, but that doesn’t mean that your RNA actually forms more structures,
just that you can resolve more subtle structural differences.</p></li>
<li><p>The Number of Unique Bit Vectors is the number of reads that were used for
clustering; it should be about 20,000 at minimum, and ideally ≥ 30,000.
If you have &lt; 20,000 unique bit vectors, then clustering will probably not
be able to find real clusters; so if the Optimal Number of Clusters is 1,
then that does not mean your RNA necessarily forms only one structure.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Expectation%E2%80%93maximization_algorithm">Expectation-maximization</a> is guaranteed to find a local optimum, but not a
global optimum.
SEISMIC-RNA thus runs multiple trajectories from different starting points; if
the trajectories converge to the same solution, then that solution is likely
(but still not necessarily) the global optimum.
You must check if your trajectories converged to the same solution by checking
the fields “NRMSD from Run 0” and “Correlation with Run 0” in the report.
If all runs converged to identical solutions, then every NRMSD would be 0 and
every Correlation would be 1.
Generally, the runs are sufficiently reproducible if runs 1 and 2 have NRMSDs
less than 0.05 and Correlations greater than 0.98 with respect to run 0.
If not, then there you have no evidence that run 0 is the global optimum for
that number of clusters, so it would be best to rerun clustering using more
independent runs to increase the chances of finding the global optimum.</p></li>
</ul>
</section>
<section id="cluster-troubleshoot-and-optimize">
<h2>Cluster: Troubleshoot and optimize<a class="headerlink" href="#cluster-troubleshoot-and-optimize" title="Permalink to this headline"></a></h2>
<section id="run-cluster-with-higher-orders-without-repeating-the-work-already-done">
<h3>Run Cluster with higher orders, without repeating the work already done<a class="headerlink" href="#run-cluster-with-higher-orders-without-repeating-the-work-already-done" title="Permalink to this headline"></a></h3>
<p>The tool <code class="docutils literal notranslate"><span class="pre">+addclust</span></code> exists for this purpose: see <a class="reference internal" href="../adjclust.html#addclust"><span class="std std-ref">Command line for adding orders</span></a>.</p>
</section>
<section id="delete-unnecessary-higher-orders-without-repeating-the-work-already-done">
<h3>Delete unnecessary higher orders, without repeating the work already done<a class="headerlink" href="#delete-unnecessary-higher-orders-without-repeating-the-work-already-done" title="Permalink to this headline"></a></h3>
<p>The tool <code class="docutils literal notranslate"><span class="pre">+delclust</span></code> exists for this purpose: see <a class="reference internal" href="../adjclust.html#delclust"><span class="std std-ref">Command line for deleting orders</span></a>.</p>
</section>
<section id="cluster-takes-too-long-to-finish">
<h3>Cluster takes too long to finish<a class="headerlink" href="#cluster-takes-too-long-to-finish" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Adjust the settings of <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">cluster</span></code>:</p>
<ul>
<li><p>Increase the threshold for convergence (<code class="docutils literal notranslate"><span class="pre">--em-thresh</span></code>).
Larger thresholds will make clustering converge in fewer iterations at the
cost of making the runs end at more variable solutions.
Check the Log Likelihood per Run field to verify that clustering is finding
the global optimum; see <a class="reference internal" href="#clust-verify"><span class="std std-ref">Cluster: Verify clusters</span></a> for more information.</p></li>
<li><p>Decrease the number of independent runs (<code class="docutils literal notranslate"><span class="pre">--em-runs</span></code>/<code class="docutils literal notranslate"><span class="pre">-e</span></code>) to 3 or 4;
don’t go below 3 for anything you intend to publish, or else you won’t be
able to tell if your clustering is finding the global optimum.</p></li>
</ul>
</li>
</ul>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="mask.html" class="btn btn-neutral float-left" title="Mask: Define mutations and sections to filter reads and positions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="join.html" class="btn btn-neutral float-right" title="Join: Merge sections (horizontally) from the Mask or Cluster step" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>