<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Graph: Plot data from tables and/or structures and compare samples &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Export: Export a file of each sample for the seismic-graph web app" href="export.html" />
    <link rel="prev" title="Fold: Predict RNA secondary structures using mutation rates" href="fold.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="relate.html">Relate: Compute relationships between references and aligned reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Merge samples (vertically) from the Relate step</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="join.html">Join: Merge sections (horizontally) from the Mask or Cluster step</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Graph: Plot data from tables and/or structures and compare samples</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#graph-types-of-graphs">Graph: Types of graphs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#graph-default-graphs">Graph: Default graphs</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lists.html">List Positions Matching Criteria</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adjclust.html">Add/Delete Orders to/from an Already-Clustered Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Graph: Plot data from tables and/or structures and compare samples</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/graph.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="graph-plot-data-from-tables-and-or-structures-and-compare-samples">
<h1>Graph: Plot data from tables and/or structures and compare samples<a class="headerlink" href="#graph-plot-data-from-tables-and-or-structures-and-compare-samples" title="Permalink to this headline"></a></h1>
<section id="graph-types-of-graphs">
<h2>Graph: Types of graphs<a class="headerlink" href="#graph-types-of-graphs" title="Permalink to this headline"></a></h2>
<p>SEISMIC-RNA can create eight types of graphs:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 9%" />
<col style="width: 91%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Graph</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>profile</p></td>
<td><p>Bar graph of relationships(s) (e.g. mutation rate or read coverage) per position</p></td>
</tr>
<tr class="row-odd"><td><p>delprof</p></td>
<td><p>Bar graph of differences between two profiles per position</p></td>
</tr>
<tr class="row-even"><td><p>scatter</p></td>
<td><p>Scatter plot comparing two profiles</p></td>
</tr>
<tr class="row-odd"><td><p>corroll</p></td>
<td><p>Rolling correlation/comparison of two profiles</p></td>
</tr>
<tr class="row-even"><td><p>histpos</p></td>
<td><p>Histogram of relationship(s) per position</p></td>
</tr>
<tr class="row-odd"><td><p>histread</p></td>
<td><p>Histogram of relationship(s) per read</p></td>
</tr>
<tr class="row-even"><td><p>roc</p></td>
<td><p>Receiver operating characteristic (ROC) curve comparing a profile to a structure</p></td>
</tr>
<tr class="row-odd"><td><p>aucroll</p></td>
<td><p>Rolling area under the ROC curve (AUC-ROC) comparing a profile to a structure</p></td>
</tr>
</tbody>
</table>
<p>See <a class="reference internal" href="../graph/index.html"><span class="doc">Graph Results</span></a> for details and instructions.</p>
</section>
<section id="graph-default-graphs">
<h2>Graph: Default graphs<a class="headerlink" href="#graph-default-graphs" title="Permalink to this headline"></a></h2>
<p>You can make six graphs when you run the entire workflow (<code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>):</p>
<ul class="simple">
<li><p>Profile of the fraction of reads mutated (the mutation rate) at each position.</p></li>
<li><p>Profile of the fraction of each type of mutation at each position.</p></li>
<li><p>Profile of the number of reads with information at each position.</p></li>
<li><p>Histogram of the number of positions mutated in each read.</p></li>
<li><p>ROC for each predicted structure (with <code class="docutils literal notranslate"><span class="pre">--fold</span></code>).</p></li>
<li><p>Rolling AUC-ROC for each predicted structure (with <code class="docutils literal notranslate"><span class="pre">--fold</span></code>).</p></li>
</ul>
<p>These graphs are created by default in <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>.
To make additional types of graphs, see <a class="reference internal" href="../graph/index.html"><span class="doc">Graph Results</span></a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="fold.html" class="btn btn-neutral float-left" title="Fold: Predict RNA secondary structures using mutation rates" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="export.html" class="btn btn-neutral float-right" title="Export: Export a file of each sample for the seismic-graph web app" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>