<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Relate: Compute relationships between references and aligned reads &mdash; seismic-rna  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/favicon-32x32.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Pool: Merge samples (vertically) from the Relate step" href="pool.html" />
    <link rel="prev" title="Align: Trim FASTQ files and align them to reference sequences" href="align.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seismic-rna
              <img src="../../_static/logo-200.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../steps/index.html">Step-By-Step</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">How To</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Run SEISMIC-RNA</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="demult.html">Demultiplex: Split multiplexed FASTQ files by their barcodes</a></li>
<li class="toctree-l3"><a class="reference internal" href="align.html">Align: Trim FASTQ files and align them to reference sequences</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Relate: Compute relationships between references and aligned reads</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#relate-input-files">Relate: Input files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#relate-settings">Relate: Settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#relate-output-files">Relate: Output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#relate-troubleshoot-and-optimize">Relate: Troubleshoot and optimize</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pool.html">Pool: Merge samples (vertically) from the Relate step</a></li>
<li class="toctree-l3"><a class="reference internal" href="mask.html">Mask: Define mutations and sections to filter reads and positions</a></li>
<li class="toctree-l3"><a class="reference internal" href="cluster.html">Cluster: Infer alternative structures by clustering reads’ mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="join.html">Join: Merge sections (horizontally) from the Mask or Cluster step</a></li>
<li class="toctree-l3"><a class="reference internal" href="table.html">Table: Count mutations for each read and position</a></li>
<li class="toctree-l3"><a class="reference internal" href="fold.html">Fold: Predict RNA secondary structures using mutation rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="graph.html">Graph: Plot data from tables and/or structures and compare samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="export.html">Export: Export a file of each sample for the seismic-graph web app</a></li>
<li class="toctree-l3"><a class="reference internal" href="wf.html">Workflow: Run all steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../graph/index.html">Graph Results</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lists.html">List Positions Matching Criteria</a></li>
<li class="toctree-l2"><a class="reference internal" href="../adjclust.html">Add/Delete Orders to/from an Already-Clustered Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cleanfa.html">Clean FASTA Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inputs.html">List Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sections.html">Define Sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalize.html">Normalize Mutation Rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parallel.html">Parallelize Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="../logging.html">Log Messages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Commands, Arguments, Options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/seismicrna.html">seismicrna package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../formats/index.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data/index.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../algos/index.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../issues.html">Bugs and Requests</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../writeme.html">How to Write this Manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seismic-rna</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">How To</a></li>
          <li class="breadcrumb-item"><a href="index.html">Run SEISMIC-RNA</a></li>
      <li class="breadcrumb-item active">Relate: Compute relationships between references and aligned reads</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/howto/run/relate.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="relate-compute-relationships-between-references-and-aligned-reads">
<h1>Relate: Compute relationships between references and aligned reads<a class="headerlink" href="#relate-compute-relationships-between-references-and-aligned-reads" title="Permalink to this headline"></a></h1>
<section id="relate-input-files">
<h2>Relate: Input files<a class="headerlink" href="#relate-input-files" title="Permalink to this headline"></a></h2>
<section id="relate-input-file-reference-sequences">
<span id="relate-refs"></span><h3>Relate input file: Reference sequences<a class="headerlink" href="#relate-input-file-reference-sequences" title="Permalink to this headline"></a></h3>
<p>You need one file of reference sequences in FASTA format (for details on this
format, see <a class="reference internal" href="../../formats/data/fasta.html"><span class="doc">FASTA: Reference sequences</span></a>).
If your file has characters or formatting incompatible with SEISMIC-RNA, then
you can fix it using the <a class="reference internal" href="../cleanfa.html"><span class="doc">Clean FASTA Files</span></a> tool.</p>
</section>
<section id="relate-input-file-alignment-maps">
<h3>Relate input file: Alignment maps<a class="headerlink" href="#relate-input-file-alignment-maps" title="Permalink to this headline"></a></h3>
<p>You can give any number of alignment map files, each of which must be in SAM,
BAM, or CRAM (collectively, “XAM”) format.
See <a class="reference internal" href="../../formats/data/xam.html"><span class="doc">SAM, BAM, and CRAM: Alignment Maps</span></a> for more information.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The references in the FASTA file must match those to which the reads in the
alignment map were aligned.
Discrepancies can cause the Relate step to fail or produce erroneous output.
You can assume that the references match if you use the same (unmodified)
FASTA file for both the <code class="docutils literal notranslate"><span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">relate</span></code> commands, or if you run
both steps using the command <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>.</p>
</div>
<p>Provide the alignment map files as a list after the FASTA file.
See <a class="reference internal" href="../inputs.html"><span class="doc">List Input Files</span></a> for ways to list multiple files.
For example, to compute relation vectors for reads from <code class="docutils literal notranslate"><span class="pre">sample-1</span></code> aligned to
references <code class="docutils literal notranslate"><span class="pre">ref-1</span></code> and <code class="docutils literal notranslate"><span class="pre">ref-2</span></code>, and from <code class="docutils literal notranslate"><span class="pre">sample-2</span></code> aligned to reference
<code class="docutils literal notranslate"><span class="pre">ref-1</span></code>, use the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seismic</span> <span class="n">relate</span> <span class="p">{</span><span class="n">refs</span><span class="o">.</span><span class="n">fa</span><span class="p">}</span> <span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mf">1.</span><span class="n">cram</span> <span class="n">sample</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mf">2.</span><span class="n">cram</span> <span class="n">sample</span><span class="o">-</span><span class="mi">2</span><span class="o">/</span><span class="n">align</span><span class="o">/</span><span class="n">ref</span><span class="o">-</span><span class="mf">1.</span><span class="n">cram</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">{refs.fa}</span></code> is the path to the file of reference sequences.</p>
</section>
</section>
<section id="relate-settings">
<h2>Relate: Settings<a class="headerlink" href="#relate-settings" title="Permalink to this headline"></a></h2>
<section id="relate-settings-shared-with-alignment">
<h3>Relate settings shared with alignment<a class="headerlink" href="#relate-settings-shared-with-alignment" title="Permalink to this headline"></a></h3>
<p>Because you can begin the SEISMIC-RNA workflow at <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> or, if you
already have alignment map files, can begin at <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code>, these two
commands share the options <code class="docutils literal notranslate"><span class="pre">--phred-enc</span></code>, <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>, <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code>, and
<code class="docutils literal notranslate"><span class="pre">--out-dir</span></code>.</p>
</section>
<section id="relate-setting-minimum-phred-score">
<h3>Relate setting: Minimum Phred score<a class="headerlink" href="#relate-setting-minimum-phred-score" title="Permalink to this headline"></a></h3>
<p>In the Relate step, you can flag bases with low quality scores as ambiguous, as
if they were <code class="docutils literal notranslate"><span class="pre">N</span></code>s.
This step serves a purpose similar to that of quality trimming during the Align
step (see <a class="reference internal" href="align.html#quality-trimming"><span class="std std-ref">How to trim low-quality base calls</span></a>).
The difference is that quality trimming removes low-quality bases by shortening
reads from their ends, while the minimum quality score in the Relate step flags
low-quality bases located anywhere in the reads, while preserving read lengths.
See <a class="reference internal" href="../../data/relate/codes.html#relate-low-qual"><span class="std std-ref">Encoding low-quality base calls</span></a> for a more detailed description of how this works.</p>
<p>To set the minimum quality score, use <code class="docutils literal notranslate"><span class="pre">--min-phred</span></code>.
The default is 25, meaning that base calls with a probabilities of at least
10<sup>-2.5</sup> = 0.3% of being incorrect are flagged as ambiguous.
(See <a class="reference internal" href="../../formats/data/fastq.html#phred-encodings"><span class="std std-ref">Phred quality score encodings</span></a> for an explanation of quality scores.)
For example, if a <code class="docutils literal notranslate"><span class="pre">T</span></code> is called as a match with a quality score of 20, then it
would be flagged as possibly a match and possibly a subsitution to A, C, or G.</p>
</section>
<section id="relate-setting-ambiguous-insertions-and-deletions">
<h3>Relate setting: Ambiguous insertions and deletions<a class="headerlink" href="#relate-setting-ambiguous-insertions-and-deletions" title="Permalink to this headline"></a></h3>
<p>When insertions and deletions (indels) occur in repetitive regions, determining
which base(s) were inserted or deleted can be impossible due to the repetitive
reference sequence itself, even if the reads were perfectly free of errors.
To handle ambiguous indels, SEISMIC-RNA introduces a new algorithm that finds
all possible indels that could have produced the observed read (for details on
this algorithm, see <span class="xref std std-doc">../algos/ambrel</span>).
This algorithm is enabled by default.
If you do not need to identify ambiguous indels, then you can disable this
algorithm with <code class="docutils literal notranslate"><span class="pre">--no-ambrel</span></code>, which will speed up the Relate step at the cost
of reducing its accuracy on indels.</p>
</section>
<section id="relate-setting-batch-size">
<h3>Relate setting: Batch size<a class="headerlink" href="#relate-setting-batch-size" title="Permalink to this headline"></a></h3>
<p>In the Relate step, you can divide up your data into batches to speed up the
analysis and reduce the amount of memory needed.
For an explanation of batching and how to use it, see <a class="reference internal" href="../parallel.html#batches"><span class="std std-ref">Batches and Benefits</span></a>.
You can specify batch size (in millions of base calls) using <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code>,
which is <code class="docutils literal notranslate"><span class="pre">64.0</span></code> (64 million base calls) by default.
Relate uses the batch size to calculate the number of reads in each batch.
The number of relationship bytes per batch, <em>B</em>, is the number of relationship
bytes per read, <em>L</em>, times the number of reads per batch, <em>N</em>:</p>
<p><em>B</em> = <em>LN</em></p>
<p>Since <em>L</em> is the length of the reference sequence and <em>B</em> is <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code>:</p>
<p><em>N</em> = <em>B</em>/<em>L</em></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>SEISMIC-RNA will aim to put exactly <em>N</em> reads in each batch but the last
(the last batch can be smaller because it has just the leftover reads).
If the reads are single-ended or were not aligned in <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a>, then
every batch but the last will contain exactly <em>N</em> reads.
If the reads are paired-ended and were aligned in <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#mixed-mode-paired-where-possible-unpaired-otherwise">mixed mode</a>, then
batches may contain more than <em>N</em> reads, up to a maximum of 2<em>N</em> in the
extreme case that only one read aligned in every mate pair.</p>
</div>
</section>
<section id="relate-setting-overhangs">
<h3>Relate setting: Overhangs<a class="headerlink" href="#relate-setting-overhangs" title="Permalink to this headline"></a></h3>
<p>When relating sequencing reads, SEISMIC-RNA will by default compute
relationships for all base calls between the smallest 5’ aligned position and
the greatest 3’ aligned position.
This occurs independent of mate alignment orientation, and can result in the
relating of base calls that fall outside the region between mate starts, for
instance, if the 5’ mate aligns to the negative strand at a position less than
the 3’ mate start position on the positive strand.
In certain rare circumstances, like when adapter trimming is inconsistent, or
when using randomized adapter sequences during library preparation, this can
result in SEISMIC-RNA calculating relationships for extraneous extensions.
The default behavior <code class="docutils literal notranslate"><span class="pre">--overhangs</span></code> can be disabled in favor of a more
conservative approach <code class="docutils literal notranslate"><span class="pre">--no-overhangs</span></code>, where only base calls greater than
the 5’ mate start and less than the 3’ mate start positions
(i.e. within the insert) are related.</p>
</section>
</section>
<section id="relate-output-files">
<h2>Relate: Output files<a class="headerlink" href="#relate-output-files" title="Permalink to this headline"></a></h2>
<p>All output files go into the directory <code class="docutils literal notranslate"><span class="pre">{out}/{sample}/relate/{ref}</span></code>, where
<code class="docutils literal notranslate"><span class="pre">{out}</span></code> is the output directory, <code class="docutils literal notranslate"><span class="pre">{sample}</span></code> is the sample, and <code class="docutils literal notranslate"><span class="pre">{ref}</span></code> is
the name of the reference.</p>
<section id="relate-output-file-batch-of-relation-vectors">
<h3>Relate output file: Batch of relation vectors<a class="headerlink" href="#relate-output-file-batch-of-relation-vectors" title="Permalink to this headline"></a></h3>
<p>Each batch of relation vectors contains a <code class="docutils literal notranslate"><span class="pre">RelateBatchIO</span></code> object and is saved
to the file <code class="docutils literal notranslate"><span class="pre">relate-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <a class="reference internal" href="../../data/relate/relate.html"><span class="doc">Relate Batch</span></a> for details on the data structure.
See <a class="reference internal" href="../../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for more information on brickle files.</p>
</section>
<section id="relate-output-file-batch-of-read-names">
<h3>Relate output file: Batch of read names<a class="headerlink" href="#relate-output-file-batch-of-read-names" title="Permalink to this headline"></a></h3>
<p>Within each batch, the relate step assigns an index (a nonnegative integer) to
each read and writes a file mapping the indexes to the read names.
Each batch of read names contains a <code class="docutils literal notranslate"><span class="pre">QnamesBatchIO</span></code> object and is saved to the
file <code class="docutils literal notranslate"><span class="pre">qnames-batch-{num}.brickle</span></code>, where <code class="docutils literal notranslate"><span class="pre">{num}</span></code> is the batch number.
See <a class="reference internal" href="../../data/relate/qnames.html"><span class="doc">Read Names Batch</span></a> for details on the data structure.
See <a class="reference internal" href="../../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for more information on brickle files.</p>
</section>
<section id="relate-output-file-reference-sequence">
<h3>Relate output file: Reference sequence<a class="headerlink" href="#relate-output-file-reference-sequence" title="Permalink to this headline"></a></h3>
<p>The relate step writes the reference sequence as a <code class="docutils literal notranslate"><span class="pre">RefseqIO</span></code> object to the
file <code class="docutils literal notranslate"><span class="pre">refseq.brickle</span></code>.
See <a class="reference internal" href="../../data/relate/refseq.html"><span class="doc">Reference Sequence</span></a> for details on the data structure.
See <a class="reference internal" href="../../formats/data/brickle.html"><span class="doc">Brickle: Compressed Python Objects</span></a> for more information on brickle files.</p>
</section>
<section id="relate-output-file-relate-report">
<h3>Relate output file: Relate report<a class="headerlink" href="#relate-output-file-relate-report" title="Permalink to this headline"></a></h3>
<p>SEISMIC-RNA also writes a report file, <code class="docutils literal notranslate"><span class="pre">relate-report.json</span></code>, that records the
settings you used for running the Relate step and summarizes the results.
See <a class="reference internal" href="../../formats/report/relate.html"><span class="doc">Relate Report</span></a> for more information.</p>
</section>
</section>
<section id="relate-troubleshoot-and-optimize">
<h2>Relate: Troubleshoot and optimize<a class="headerlink" href="#relate-troubleshoot-and-optimize" title="Permalink to this headline"></a></h2>
<p>If you encounted errors during the Relate step, then the most likely cause is
that the FASTA file or settings you used for the Relate step differ from those
that you used during alignment.</p>
<section id="insufficient-reads-in-file">
<h3>Insufficient reads in {file} …<a class="headerlink" href="#insufficient-reads-in-file" title="Permalink to this headline"></a></h3>
<p>This error means that you provided a SAM/BAM/CRAM file containing fewer reads
than the minimum number set by <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code> (<code class="docutils literal notranslate"><span class="pre">-N</span></code>).
There are two common causes of this error:</p>
<ul class="simple">
<li><p>You ran <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code> separately (instead of with
<code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>), and you used a larger value for <code class="docutils literal notranslate"><span class="pre">--min-reads</span></code> during the
Relate step than the Align step.
To check if this happened, open your report files from Align and Relate and
see if the field “Minimum number of reads in an alignment map” has a larger
value in the Relate report.</p></li>
<li><p>You ran alignment outside of SEISMIC-RNA or obtained alignment map files from
an external source, and some of the alignment maps have insufficient reads.</p></li>
</ul>
<p>The solution for the problem is to ensure that you run <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code> with
<code class="docutils literal notranslate"><span class="pre">--min-reads</span></code> set to the minimum number of reads you actually want during the
Relate step.
As long as you do so, you may ignore error messages about insufficient reads,
since these messages just indicate that SEISMIC-RNA is skipping alignment maps
with insufficient reads, which is exactly what you want to happen.</p>
</section>
<section id="read-read-mapped-with-a-quality-score-score">
<h3>Read {read} mapped with a quality score {score} …<a class="headerlink" href="#read-read-mapped-with-a-quality-score-score" title="Permalink to this headline"></a></h3>
<p>This error means that a read inside an alignment file aligned with a mapping
quality lower than the minimum set by <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code>.
There are two common causes of this error:</p>
<ul class="simple">
<li><p>You ran <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> and <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code> separately (instead of with
<code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>), and you used a larger value for <code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code> during the
Relate step than the Align step.
To check if this happened, open your report files from Align and Relate and
see if the field “Minimum mapping quality to use an aligned read” has a larger
value in the Relate report.</p></li>
<li><p>You ran alignment outside of SEISMIC-RNA or obtained alignment map files from
an external source, and some reads in the alignment maps have insufficient
mapping quality.</p></li>
</ul>
<p>The solution for the problem is to ensure that you run <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">relate</span></code> with
<code class="docutils literal notranslate"><span class="pre">--min-mapq</span></code> set to the minimum mapping quality you actually want during the
Relate step.
As long as you do so, you may ignore error messages about insufficient quality,
since these messages just indicate that SEISMIC-RNA is skipping reads with
with insufficient mapping quality, which is exactly what you want to happen.</p>
</section>
<section id="read-read-mapped-to-a-reference-named-name">
<h3>Read {read} mapped to a reference named {name} …<a class="headerlink" href="#read-read-mapped-to-a-reference-named-name" title="Permalink to this headline"></a></h3>
<p>This error means that a read inside an alignment file aligned to a reference
whose name does not match the name of the alignment file (minus the extension).
For example, if your alignment map file <code class="docutils literal notranslate"><span class="pre">azure.cram</span></code> contains a read that
aligned to a reference named <code class="docutils literal notranslate"><span class="pre">cyan</span></code> (instead of <code class="docutils literal notranslate"><span class="pre">azure</span></code>), then you will get
this error message.</p>
<p>If you aligned the reads using <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">align</span></code> or <code class="docutils literal notranslate"><span class="pre">seismic</span> <span class="pre">wf</span></code>, then this
error should never occur (unless you renamed or modified the output files).
Otherwise, you can solve the problem by ensuring that</p>
<ul class="simple">
<li><p>Each alignment map file contains reads that aligned to only one reference.</p></li>
<li><p>Each alignment map file is named (up to the file extension) the same as the
one reference to which all of the reads aligned.</p></li>
</ul>
</section>
<section id="relate-crashes-or-hangs-while-producing-few-or-no-batch-files">
<h3>Relate crashes or hangs while producing few or no batch files<a class="headerlink" href="#relate-crashes-or-hangs-while-producing-few-or-no-batch-files" title="Permalink to this headline"></a></h3>
<p>Most likely, your system has run out of memory.
You can confirm using a program that monitors memory usage (such as <code class="docutils literal notranslate"><span class="pre">top</span></code> in a
Linux/macOS terminal, Activity Monitor on macOS, or Task Manager on Windows).
If so, then rerun Relate with adjustments to one or both settings:</p>
<ul class="simple">
<li><p>Use smaller batches (with <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code>) to limit the size of each batch,
at the cost of having more files with a larger total size.</p></li>
<li><p>Use fewer processors (with <code class="docutils literal notranslate"><span class="pre">--max-procs</span></code>) to limit the memory usage, at the
cost of slower processing.</p></li>
</ul>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="align.html" class="btn btn-neutral float-left" title="Align: Trim FASTQ files and align them to reference sequences" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pool.html" class="btn btn-neutral float-right" title="Pool: Merge samples (vertically) from the Relate step" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, the Rouskin Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>